test/unit/MatchingEngine.spec.ts

```diff
@@ -38,59 +44,143 @@ describe('MatchingEngine.getMatchingQuantity', () => {
+    );
+    expect(res).to.be.true;
+  });
+
+  it('should prioritize earlier createdAt when prices are equal on DESC ordering direction', () => {
+    const comparator = MatchingEngine.getOrdersPriorityQueueComparator(enums.orderingDirections.DESC);
+    const res = comparator(
+      createOrder(5, 10, new Date(1)),
+      createOrder(5, -10, new Date(2)),
+    );
+    expect(res).to.be.true;
+  });
+});
+
+describe('MatchingEngine.splitOrderByQuantity', () => {
+  it('should split buy orders properly', () => {
+    const orderQuantity = 10;
+    const targetQuantity = 6;
+    const { target, remaining } = MatchingEngine.splitOrderByQuantity(
+      createOrder(5, orderQuantity),
+      targetQuantity,
+    );
+    expect(target.quantity).to.be.equal(targetQuantity);
+    expect(remaining.quantity).to.be.equal(orderQuantity - targetQuantity);
+  });
+
+  it('should split sell orders properly', () => {
+    const orderQuantity = -10;
+    const targetQuantity = 4;
+    const { target, remaining } = MatchingEngine.splitOrderByQuantity(
+      createOrder(5, orderQuantity),
+      targetQuantity,
+    );
+    expect(target.quantity).to.be.equal(targetQuantity * -1);
+    expect(remaining.quantity).to.be.equal(orderQuantity + targetQuantity);
+  });
+
+  it('should not work when targetQuantity higher than quantity of order', () => {
+    expect(() => MatchingEngine.splitOrderByQuantity(
+      createOrder(5, 5),
+      10,
+    )).to.throw('order abs quantity should be higher than targetQuantity');
+  });
+});
+
+describe('MatchingEngine.match', () => {
+  it('should fully match with two maker orders', () => {
+    const engine = new MatchingEngine(pairId, true, [], [
+      createDbOrder(5, -5),
+      createDbOrder(5, -5),
+    ], [], []);
+    const matchAgainst = [engine.priorityQueues.peerSellOrders];
+    const { remainingOrder } = MatchingEngine.match(
+      createOrder(5, 10),
+      matchAgainst,
+    );
+    expect(remainingOrder).to.be.equal(null);
+  });
+
+  it('should split taker order when makers are insufficient', () => {
+    const engine = new MatchingEngine(pairId, true, [], [
+      createDbOrder(5, -5),
+      createDbOrder(5, -4),
+    ], [], []);
+    const matchAgainst = [engine.priorityQueues.peerSellOrders];
+    const { remainingOrder } = MatchingEngine.match(
+      createOrder(5, 10),
+      matchAgainst,
+    );
+    expect(remainingOrder.quantity).to.be.equal(1);
+  });
+
+  it('should split one maker order when taker is insufficient', () => {
+    const engine = new MatchingEngine(pairId, true, [], [
+      createDbOrder(5, -5),
+      createDbOrder(5, -6),
+    ], [], []);
+    const matchAgainst = [engine.priorityQueues.peerSellOrders];
+    const { matches, remainingOrder } = MatchingEngine.match(
+      createOrder(5, 10),
+      matchAgainst,
+    );
+    expect(remainingOrder).to.be.equal(null);
+    matches.forEach((match) => {
+      expect(match.maker.quantity).to.be.equal(-5);
```

Comment:
Maybe also add a check here that the remaining sell order has -1 quantity remaining, if possible.